woocommerce/woocommerce-ios
# Fix ghost layers in the shipping class selector

Fixes #1645

## Cause of issue

After some debugging, the ghost layer doesn't look as expected (see screenshots in #1645) because the cell `BasicTableViewCell`'s layout is frame-based instead of Auto Layout based and thus the ghost layer does not adjust to its superview's frame properly. This PR updated the use of `BasicTableViewCell` to `WooBasicTableViewCell`, an Auto Layout version of `BasicTableViewCell`, with its top margin constraint configurable for other existing usage.

## Changes

- Updated `WooBasicTableViewCell` xib: replaced bottom margin constraint with vertically center constraint, and added an outlet for the top margin constraint with 0 by default to make the vertical constraint configurable.
- Extracted `WooBasicTableViewCell` styling to 3 extension helpers:
  - Plain text style (e.g. used in Order Details > View Billing Information)
  - Actionable style (e.g. used in Product Details > view product link)
  - List selector style, used in the shipping class list selector cells

## Testing

Prerequisite: have a store without shipping classes, or hard code to always show the syncing state

- Launch the app
- Go to Products tab
- Tap on a Product
- Tap "Edit"
- Tap on the shipping settings row
- Tap the Shipping class row --> the ghost layer shouldn't look weird and should be flashing

## Example screenshots

Syncing | List selector | Order Details (plain text style) | Product Details (actionable style)
-- | -- | -- | --
![Simulator Screen Shot - iPhone 11 - 2019-12-26 at 14 41 37](https://user-images.githubusercontent.com/1945542/71462842-980edf80-27ef-11ea-9aab-9f624f5776c3.png) | ![Simulator Screen Shot - iPhone 11 - 2019-12-26 at 14 42 34](https://user-images.githubusercontent.com/1945542/71462843-980edf80-27ef-11ea-9b0a-5b69f2d46558.png) | ![Simulator Screen Shot - iPhone 11 - 2019-12-26 at 14 42 53](https://user-images.githubusercontent.com/1945542/71462844-980edf80-27ef-11ea-8c3a-b2e0d88bb08b.png) | ![Simulator Screen Shot - iPhone 11 - 2019-12-26 at 14 54 14](https://user-images.githubusercontent.com/1945542/71462845-98a77600-27ef-11ea-8553-b60d0f0b90b5.png)


Update release notes:
